// @aws-cdk/aws-kinesis/lib/kinesis-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.KinesisMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class KinesisMetrics {
    static readProvisionedThroughputExceededAverage(dimensions) {
        return {
            namespace: 'AWS/Kinesis',
            metricName: 'ReadProvisionedThroughputExceeded',
            dimensions,
            statistic: 'Average',
        };
    }
    static writeProvisionedThroughputExceededAverage(dimensions) {
        return {
            namespace: 'AWS/Kinesis',
            metricName: 'WriteProvisionedThroughputExceeded',
            dimensions,
            statistic: 'Average',
        };
    }
    static getRecordsIteratorAgeMillisecondsMaximum(dimensions) {
        return {
            namespace: 'AWS/Kinesis',
            metricName: 'GetRecords.IteratorAgeMilliseconds',
            dimensions,
            statistic: 'Maximum',
        };
    }
    static putRecordSuccessAverage(dimensions) {
        return {
            namespace: 'AWS/Kinesis',
            metricName: 'PutRecord.Success',
            dimensions,
            statistic: 'Average',
        };
    }
    static putRecordsBytesAverage(dimensions) {
        return {
            namespace: 'AWS/Kinesis',
            metricName: 'PutRecords.Bytes',
            dimensions,
            statistic: 'Average',
        };
    }
    static getRecordSuccessAverage(dimensions) {
        return {
            namespace: 'AWS/Kinesis',
            metricName: 'GetRecord.Success',
            dimensions,
            statistic: 'Average',
        };
    }
    static getRecordBytesAverage(dimensions) {
        return {
            namespace: 'AWS/Kinesis',
            metricName: 'GetRecord.Bytes',
            dimensions,
            statistic: 'Average',
        };
    }
    static getRecordRecordsAverage(dimensions) {
        return {
            namespace: 'AWS/Kinesis',
            metricName: 'GetRecord.Records',
            dimensions,
            statistic: 'Average',
        };
    }
    static getRecordLatencyMaximum(dimensions) {
        return {
            namespace: 'AWS/Kinesis',
            metricName: 'GetRecord.Latency',
            dimensions,
            statistic: 'Maximum',
        };
    }
    static incomingBytesAverage(dimensions) {
        return {
            namespace: 'AWS/Kinesis',
            metricName: 'IncomingBytes',
            dimensions,
            statistic: 'Average',
        };
    }
    static incomingRecordsAverage(dimensions) {
        return {
            namespace: 'AWS/Kinesis',
            metricName: 'IncomingRecords',
            dimensions,
            statistic: 'Average',
        };
    }
}
exports.KinesisMetrics = KinesisMetrics;
