// @aws-cdk/aws-kinesisfirehose/lib/kinesisfirehose-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.FirehoseMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class FirehoseMetrics {
    static incomingBytesAverage(dimensions) {
        return {
            namespace: 'AWS/Firehose',
            metricName: 'IncomingBytes',
            dimensions,
            statistic: 'Average',
        };
    }
    static incomingRecordsAverage(dimensions) {
        return {
            namespace: 'AWS/Firehose',
            metricName: 'IncomingRecords',
            dimensions,
            statistic: 'Average',
        };
    }
    static backupToS3BytesAverage(dimensions) {
        return {
            namespace: 'AWS/Firehose',
            metricName: 'BackupToS3.Bytes',
            dimensions,
            statistic: 'Average',
        };
    }
    static backupToS3DataFreshnessAverage(dimensions) {
        return {
            namespace: 'AWS/Firehose',
            metricName: 'BackupToS3.DataFreshness',
            dimensions,
            statistic: 'Average',
        };
    }
    static backupToS3RecordsAverage(dimensions) {
        return {
            namespace: 'AWS/Firehose',
            metricName: 'BackupToS3.Records',
            dimensions,
            statistic: 'Average',
        };
    }
}
exports.FirehoseMetrics = FirehoseMetrics;
