// @aws-cdk/aws-ecs/lib/ecs-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ECSMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class ECSMetrics {
    static cpuUtilizationAverage(dimensions) {
        return {
            namespace: 'AWS/ECS',
            metricName: 'CPUUtilization',
            dimensions,
            statistic: 'Average',
        };
    }
    static memoryUtilizationAverage(dimensions) {
        return {
            namespace: 'AWS/ECS',
            metricName: 'MemoryUtilization',
            dimensions,
            statistic: 'Average',
        };
    }
    static cpuReservationAverage(dimensions) {
        return {
            namespace: 'AWS/ECS',
            metricName: 'CPUReservation',
            dimensions,
            statistic: 'Average',
        };
    }
    static memoryReservationAverage(dimensions) {
        return {
            namespace: 'AWS/ECS',
            metricName: 'MemoryReservation',
            dimensions,
            statistic: 'Average',
        };
    }
    static gpuReservationAverage(dimensions) {
        return {
            namespace: 'AWS/ECS',
            metricName: 'GPUReservation',
            dimensions,
            statistic: 'Average',
        };
    }
}
exports.ECSMetrics = ECSMetrics;
