// @aws-cdk/aws-stepfunctions/lib/stepfunctions-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.StatesMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class StatesMetrics {
    static executionTimeAverage(dimensions) {
        return {
            namespace: 'AWS/States',
            metricName: 'ExecutionTime',
            dimensions,
            statistic: 'Average',
        };
    }
    static executionsFailedAverage(dimensions) {
        return {
            namespace: 'AWS/States',
            metricName: 'ExecutionsFailed',
            dimensions,
            statistic: 'Average',
        };
    }
    static executionsSucceededAverage(dimensions) {
        return {
            namespace: 'AWS/States',
            metricName: 'ExecutionsSucceeded',
            dimensions,
            statistic: 'Average',
        };
    }
    static executionsThrottledAverage(dimensions) {
        return {
            namespace: 'AWS/States',
            metricName: 'ExecutionsThrottled',
            dimensions,
            statistic: 'Average',
        };
    }
    static executionsAbortedAverage(dimensions) {
        return {
            namespace: 'AWS/States',
            metricName: 'ExecutionsAborted',
            dimensions,
            statistic: 'Average',
        };
    }
    static executionsTimedOutAverage(dimensions) {
        return {
            namespace: 'AWS/States',
            metricName: 'ExecutionsTimedOut',
            dimensions,
            statistic: 'Average',
        };
    }
    static activityTimeAverage(dimensions) {
        return {
            namespace: 'AWS/States',
            metricName: 'ActivityTime',
            dimensions,
            statistic: 'Average',
        };
    }
    static activitiesSucceededSum(dimensions) {
        return {
            namespace: 'AWS/States',
            metricName: 'ActivitiesSucceeded',
            dimensions,
            statistic: 'Sum',
        };
    }
    static activitiesFailedSum(dimensions) {
        return {
            namespace: 'AWS/States',
            metricName: 'ActivitiesFailed',
            dimensions,
            statistic: 'Sum',
        };
    }
    static activitiesHeartbeatTimedOutSum(dimensions) {
        return {
            namespace: 'AWS/States',
            metricName: 'ActivitiesHeartbeatTimedOut',
            dimensions,
            statistic: 'Sum',
        };
    }
    static activitiesScheduledSum(dimensions) {
        return {
            namespace: 'AWS/States',
            metricName: 'ActivitiesScheduled',
            dimensions,
            statistic: 'Sum',
        };
    }
    static activitiesStartedSum(dimensions) {
        return {
            namespace: 'AWS/States',
            metricName: 'ActivitiesStarted',
            dimensions,
            statistic: 'Sum',
        };
    }
    static activitiesTimedOutSum(dimensions) {
        return {
            namespace: 'AWS/States',
            metricName: 'ActivitiesTimedOut',
            dimensions,
            statistic: 'Sum',
        };
    }
    static activityRunTimeAverage(dimensions) {
        return {
            namespace: 'AWS/States',
            metricName: 'ActivityRunTime',
            dimensions,
            statistic: 'Average',
        };
    }
    static activityScheduleTimeAverage(dimensions) {
        return {
            namespace: 'AWS/States',
            metricName: 'ActivityScheduleTime',
            dimensions,
            statistic: 'Average',
        };
    }
}
exports.StatesMetrics = StatesMetrics;
