// @aws-cdk/aws-elasticloadbalancing/lib/elasticloadbalancing-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ELBMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class ELBMetrics {
    static latencyAverage(dimensions) {
        return {
            namespace: 'AWS/ELB',
            metricName: 'Latency',
            dimensions,
            statistic: 'Average',
        };
    }
    static backendConnectionErrorsAverage(dimensions) {
        return {
            namespace: 'AWS/ELB',
            metricName: 'BackendConnectionErrors',
            dimensions,
            statistic: 'Average',
        };
    }
    static requestCountAverage(dimensions) {
        return {
            namespace: 'AWS/ELB',
            metricName: 'RequestCount',
            dimensions,
            statistic: 'Average',
        };
    }
    static spilloverCountAverage(dimensions) {
        return {
            namespace: 'AWS/ELB',
            metricName: 'SpilloverCount',
            dimensions,
            statistic: 'Average',
        };
    }
    static surgeQueueLengthAverage(dimensions) {
        return {
            namespace: 'AWS/ELB',
            metricName: 'SurgeQueueLength',
            dimensions,
            statistic: 'Average',
        };
    }
    static healthyHostCountAverage(dimensions) {
        return {
            namespace: 'AWS/ELB',
            metricName: 'HealthyHostCount',
            dimensions,
            statistic: 'Average',
        };
    }
    static unHealthyHostCountAverage(dimensions) {
        return {
            namespace: 'AWS/ELB',
            metricName: 'UnHealthyHostCount',
            dimensions,
            statistic: 'Average',
        };
    }
    static httpCodeBackend2XxAverage(dimensions) {
        return {
            namespace: 'AWS/ELB',
            metricName: 'HTTPCode_Backend_2XX',
            dimensions,
            statistic: 'Average',
        };
    }
    static httpCodeBackend3XxAverage(dimensions) {
        return {
            namespace: 'AWS/ELB',
            metricName: 'HTTPCode_Backend_3XX',
            dimensions,
            statistic: 'Average',
        };
    }
    static httpCodeBackend4XxAverage(dimensions) {
        return {
            namespace: 'AWS/ELB',
            metricName: 'HTTPCode_Backend_4XX',
            dimensions,
            statistic: 'Average',
        };
    }
    static httpCodeBackend5XxAverage(dimensions) {
        return {
            namespace: 'AWS/ELB',
            metricName: 'HTTPCode_Backend_5XX',
            dimensions,
            statistic: 'Average',
        };
    }
}
exports.ELBMetrics = ELBMetrics;
