// @aws-cdk/aws-cloudfront/lib/cloudfront-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudFrontMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class CloudFrontMetrics {
    static requestsSum(dimensions) {
        return {
            namespace: 'AWS/CloudFront',
            metricName: 'Requests',
            dimensions,
            statistic: 'Sum',
        };
    }
    static totalErrorRateAverage(dimensions) {
        return {
            namespace: 'AWS/CloudFront',
            metricName: 'TotalErrorRate',
            dimensions,
            statistic: 'Average',
        };
    }
    static bytesDownloadedSum(dimensions) {
        return {
            namespace: 'AWS/CloudFront',
            metricName: 'BytesDownloaded',
            dimensions,
            statistic: 'Sum',
        };
    }
    static bytesUploadedAverage(dimensions) {
        return {
            namespace: 'AWS/CloudFront',
            metricName: 'BytesUploaded',
            dimensions,
            statistic: 'Average',
        };
    }
    static _4XxErrorRateAverage(dimensions) {
        return {
            namespace: 'AWS/CloudFront',
            metricName: '4xxErrorRate',
            dimensions,
            statistic: 'Average',
        };
    }
    static _5XxErrorRateAverage(dimensions) {
        return {
            namespace: 'AWS/CloudFront',
            metricName: '5xxErrorRate',
            dimensions,
            statistic: 'Average',
        };
    }
}
exports.CloudFrontMetrics = CloudFrontMetrics;
