// @aws-cdk/aws-codebuild/lib/codebuild-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeBuildMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class CodeBuildMetrics {
    static succeededBuildsSum(dimensions) {
        return {
            namespace: 'AWS/CodeBuild',
            metricName: 'SucceededBuilds',
            dimensions,
            statistic: 'Sum',
        };
    }
    static failedBuildsSum(dimensions) {
        return {
            namespace: 'AWS/CodeBuild',
            metricName: 'FailedBuilds',
            dimensions,
            statistic: 'Sum',
        };
    }
    static buildsSum(dimensions) {
        return {
            namespace: 'AWS/CodeBuild',
            metricName: 'Builds',
            dimensions,
            statistic: 'Sum',
        };
    }
    static durationAverage(dimensions) {
        return {
            namespace: 'AWS/CodeBuild',
            metricName: 'Duration',
            dimensions,
            statistic: 'Average',
        };
    }
}
exports.CodeBuildMetrics = CodeBuildMetrics;
